### ISSUE TITLE
Allow overloading binary operations between jax.Array and custom array-like types (perhaps via `__array_priority__`)

### ISSUE THREAD

**Issue by danieldjohnson** [google/jax#22240]
For downstream libraries built on top of JAX, it is sometimes useful to define "wrapped arrays" that can be converted to JAX arrays. For instance, [Penzai's named array system](https://penzai.readthedocs.io/en/stable/notebooks/named_axes.html) defines a `NamedArray` type, which allows top-level arrays to have optional axis names.

Ideally, it would be nice if such wrapped arrays could both:

- Be automatically convertable to JAX arrays using `jax.array(my_wrapped_array)`
- Override binary operations, so that `my_jax_array + my_wrapped_array` and `my_wrapped_array + my_jax_array` both call `MyWrappedArray.__add__` or `MyWrappedArray.__radd__`.

However, this is currently not possible in JAX due to interactions with Python's operator overloading behavior.

- When evaluating `lhs + rhs`, Python will only call `rhs.__radd__(lhs)` if `lhs.__add__(rhs)` returns `NotImplemented`, or if `type(rhs)` is a subclass of `type(lhs)` (described [here](https://docs.python.org/3/reference/datamodel.html#object.__radd__)).
- JAX allows custom types to be converted to JAX arrays by calling their `__jax_array__` method, if it exists.
- However, in binary operations, if `lhs` is a JAX array and `rhs` is something with a `__jax_array__` method, [JAX uses it](https://github.com/google/jax/blob/5fcd50b7faadfd14433b642c82710493c7be37a5/jax/_src/numpy/array_methods.py#L258) to convert `rhs` to a JAX array, instead of returning `NotImplemented`.

A similar interaction also arises with Numpy arrays and the `__array__` method, but Numpy circumvents it using the [`__array_priority__`](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_priority__) attribute: when evaluating a binary operation between two operations, `lhs.__add__(rhs)` will return `NotImplemented` if `lhs` is a Numpy array and `rhs` has a higher array priority. (Numpy uses this for custom array types like masked arrays.)

JAX currently *sets* the `__array_priority__` attribute (to a very large value) on JAX arrays and tracers. This means that binary operations between JAX arrays and Numpy arrays always return JAX arrays, and interact correctly with JAX tracing, even though JAX arrays define `__array__`.

However, JAX does not *check* the `__array_priority__` attribute when implementing binary operations on JAX arrays. As a result, it's not possible to similarly define higher-level array-like operations that always return `MyWrappedArray`, without disabling conversion of `MyWrappedArray` instances to JAX arrays entirely.

---

This limitation is the reason why Penzai's `NamedArray` does not currently allow conversion to JAX arrays using `__jax_array__` (see https://github.com/google-deepmind/penzai/issues/35). It would be great if there was a supported way to extend binary operations in JAX similar to how they can be extended in Numpy, perhaps also using the `__array_priority__` attribute.

**Comment by jakevdp** [comment]
> This limitation is the reason why Penzai's NamedArray does not currently allow conversion to JAX arrays using `__jax_array__`

Good! `__jax_array__` is undocumented, not broadly supported in JAX library code, virtually untested, and generally not recommended for use in any important code. It also breaks in subtle ways when combined with PyTree lowering. Seriously: I'd strongly recommend against implementing `__jax_array__` in your classes, because it will only cause you problems down the road.

With that recommendation in mind, it sounds like the `__array_priority__` mechanism is no longer relevant to JAX. What do you think?

**Comment by danieldjohnson** [comment]
I see. In that case, I agree handling array priority probably isn't relevant.

I'll close this for now. If `__jax_array__` ever becomes documented / recommended for downstream use, it might be worth revisiting.

**Comment by jakevdp** [comment]
For context, here's some of the justification for why `__jax_array__` was added in the first place: https://github.com/google/jax/pull/4725#issuecomment-746457957

It was an experiment that was not meant to be load-bearing. That's more or less unchanged until today (though as I feared in that thread, a number of downstream users have discovered it and begun depending on it, so we can't easily remove the experimental implementation at this point).
